ENH: Fix total seconds in timedelta

deponovo opened the pull request:
- [x] closes #40946
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

As of now, I could not figure out what's leading to test failing in pandas\tests\scalar\timestamp\test_constructors.py:609.

deponovo committed:
BUG: fixing `total_seconds` for timedeltas smaller then 1 microsecond

jreback reviewed (changes requested):
needs a test case

deponovo committed:
TST: added testing of `Timedelta.total_seconds`

jreback commented on pandas/tests/tslibs/test_timedeltas.py:
```diff
@@ -37,3 +37,23 @@ def test_huge_nanoseconds_overflow():
+        (Timedelta("500ns"), 5e-7),
+        (Timedelta(nanoseconds=500), 5e-7),
+        (Timedelta(seconds=1, nanoseconds=500), 1 + 5e-7),
+        # require GH45108
```
don't add comment code instead just add this case as a np.timedelta64  constructed from ns ints which preserves this (or add as a followup)

deponovo replied:
Fixed

jreback commented on pandas/_libs/tslibs/timedeltas.pyx:
```diff
@@ -1520,6 +1520,10 @@ class Timedelta(_Timedelta):
         div = other // self
         return div, other - div * self
 
+    def total_seconds(self):
```
add a return type

deponovo replied:
Done

jreback commented on pandas/_libs/tslibs/timedeltas.pyx:
```diff
@@ -1520,6 +1520,10 @@ class Timedelta(_Timedelta):
         div = other // self
         return div, other - div * self
 
+    def total_seconds(self):
```
can you add a full-docstring, e.g. need examples, wouldn't hurt to reference the python timedelta.total_seconds method as well.

deponovo replied:
Done

jreback commented:
also some tests are failing

deponovo commented:
> also some tests are failing

Yes, I referred that in the entry PR message. I could not figure out what the reason is and I am lacking the time to further seek the source. Maybe somebody already knows the issue or could help? 
The hour difference from a given timezone flips ahead (but short) of the real timestamp flip to the next hour.

mroeschke commented:
This comment might potentially be relevant https://github.com/pandas-dev/pandas/pull/31155/files#r370721655

deponovo committed:
CLN: rebased and resolved conflict

deponovo committed:
ENH: enhanced the `Timedelta.total_seconds()` override for compatibility reasons
DOC: updated the docstring of the `Timedelta.total_seconds()`
TST: updated respective testing

deponovo committed:
DOC: fix docstring empty line

jreback commented on pandas/_libs/tslibs/timedeltas.pyx:
```diff
@@ -1539,6 +1539,34 @@ class Timedelta(_Timedelta):
+        >>> td.total_seconds(ns_precision=True)
+        521403.010010012
+        """
+        if ns_precision:
```
umm what is this?

deponovo replied:
I could not find a way to prevent the problems that the higher precision causes in timezone-related-info (see link from @mroeschke) and because of that I am keeping the `super` implementation the default. However, having a `total_seconds` method unable to output nanosecond precision is not consistent. My solution for this was to override the method with this version that has a documented `kwarg` such that anybody can then know about it (possibly after a bad surprise) without having to check source code.

jreback replied:
adding a keyword like this is not going to fly. if you want to try to fix that issue then happy to incorporate here (or in a pre-cursor). that looks like the test case needs to be updated.

mroeschke replied:
Agreed. The signatures need to be kept compatible with the `datetime.timedelta` methods as `Timedelta` is a subclass

deponovo replied:
Ok. Will retry at a later time.

deponovo commented:
This is puzzling. I can't still figure out the source of the prob. I transformed the `nanosecond` value from `_Timestamp` in `pandas/_libs/tslibs/timestamps.pxd` just for test purposes into a public value and set it to 0 by default. Then tested a boundary condition:

```
>>> td = pd.Timestamp(1552211999999999999,  tz="dateutil/America/Los_Angeles")
>>> td
Timestamp('2019-03-10 01:59:59.999999-0800', tz='dateutil/US/Pacific')
>>> td.value
1552211999999999999
>>> td.nanosecond
0
>>> td.nanosecond = 999  # actually just assigning the nanosecond part of the original value
>>> td
Timestamp('2019-03-10 01:59:59.999999999-0700', tz='dateutil/US/Pacific')
>>> td.value
1552211999999999999
>>> td.nanosecond
999
```

mroeschke commented:
> @mroeschke Now back to the original method signature:
> 
> ```
>     def total_seconds(self):
>         """Total seconds in the duration."""
>         return self.value / 1_000_000_000
> ```
> 
> This is puzzling. I can't still figure out the source of the prob. I modified the `nanosecond` variable from `_Timestamp` in `pandas/_libs/tslibs/timestamps.pxd` to be public (just for testing purposes) and set it to 0 by default. Then tested a boundary condition:
> 
> ```
> >>> td = pd.Timestamp(1552211999999999999,  tz="dateutil/America/Los_Angeles")
> >>> td
> Timestamp('2019-03-10 01:59:59.999999-0800', tz='dateutil/US/Pacific')
> >>> td.value
> 1552211999999999999
> >>> td.nanosecond
> 0
> >>> td.nanosecond = 999  # actually just assigning the nanosecond part of the original value
> >>> td  # note the false -7
> Timestamp('2019-03-10 01:59:59.999999999-0700', tz='dateutil/US/Pacific')
> >>> td.value
> 1552211999999999999
> >>> td.nanosecond
> 999
> ```

Looking into https://github.com/pandas-dev/pandas/pull/31155/files#r370721655 might solve this issue

mroeschke commented:
Thanks for the pull request but appears to have gone stale. If you're interested in continuing please merge in main, address the comments and we can reopen.
